docs/library/machine.Counter.rst

```diff
@@ -0,0 +1,145 @@
+.. currentmodule:: machine
+.. _machine.Counter:
+
+class Counter -- hardware counter
+=================================
+
+A hardware counter counts the transitions or pulses on an input pin and accumulates the count into a
+hardware register. This register can be read, reset, and typically thresholds can be set to trigger
+interrupts when specific values are reached.
+
+Counters are related to timers and in some
+microprocessors use the same hardware. From a usage perspective, counters count
+edges (or pulses) that happen at arbitrary points in time and the resulting register
+values over time are expressed simply as counts whereas timers tend to count
+regular clock pulses and the resulting register values are expressed as units of time.
+
+Most hardware counter units can count up or down, can count on positive or negative
+incoming edges and can reset at/to some max value. On many microprocessors the input
+can be fed through a prescaler which divides the incoming pulses by a power of two.
+It is also common to find a filter which can cause short pulses due to glitches to be
+ignored.
+
+Example usage::
+
+    from machine import Counter, Pin
+    ctr = Counter(0, pin=4)
```

Comment:
`pin=Pin(4)`

Reply:
Yes & no... I'm copying the conversation from the slack thread https://micropython.slack.com/archives/C48TVHRQC/p1578281256187400?thread_ts=1578281225.187300&cid=C48TVHRQC:

> jimmo  1 day ago
the general idea is that anything that takes a pin should be able to accept either a machine.Pin instance or a port-specific value

> jimmo  1 day ago
so on STM32, the port-specific value is a string, which is either a "CPU" pin name (e.g. PA1) or a "BOARD" pin name (e.g. "X1" on a pyboard, or "A0", "D12", etc on an arduino-layout-compatible board)
each board definition defines a pins.csv which maps the board pins to CPU pins, and the CPU pins are generated automatically based on which stm32 part it is (edited) 

> jimmo  1 day ago
on ESP32, the port-specific value is the pin number, and the pins.csv board mapping etc isn't implemented  (i recently added basic support for board definitions, and hopefulyl someone who is more involved with ESP32 can finish it off and add pins.csv)

[...]

> TvE  1 day ago
Is it OK to use pin numbers (ints) on the esp32 now and then later the code can test the type of the pin parameter and if it's a string look up the pin by name and if it's a Pin use that?

> jimmo  1 day ago
yes, that will at least be compatible with the rest of ESP32

> jimmo  1 day ago
i hope that in the future that just a few extra calls to the esp32 version of pin_find will be all that's required

> jimmo  1 day ago
then ESP32 will support pin number, pin name, and pin obj in all cases